node_agent/na/gcp.go

```diff
@@ -0,0 +1,54 @@
+// Copyright 2017 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package na
+
+import (
+	"cloud.google.com/go/compute/metadata"
+	"github.com/golang/glog"
+	"golang.org/x/net/context"
+	"google.golang.org/grpc"
+	cred "istio.io/auth/pkg/credential"
+)
+
+type jwtAccess struct {
+	token string
+}
+
+func (j jwtAccess) GetRequestMetadata(ctx context.Context, uri ...string) (map[string]string, error) {
+	return map[string]string{
+		"authorization": j.token,
+	}, nil
+}
+
+func (j jwtAccess) RequireTransportSecurity() bool {
+	return true
+}
+
+type gcpPlatformImpl struct {
+}
+
+func (na *gcpPlatformImpl) ProperPlatform() bool {
+	return metadata.OnGCE()
+}
+
+func (na *gcpPlatformImpl) GetDialOptions(cfg *Config, fetcher cred.TokenFetcher) ([]grpc.DialOption, error) {
```

Comment:
Instead of doing it this way, can we do the following
```
type gcpPlatformImpl struct {
     fetch cred.TokenFetcher
}

func (na *gcpPlatformImpl) GetDialOptions(cfg *Config) {
    jwtKey, err := na.fetcher.FetchToken()
    ........
}
```

This way you don't need to set `fetcher` to nil conditionally as now: https://github.com/istio/auth/pull/207/files#diff-c8c2d04a54ef803c22965ceb30ff6ea3R125

And I think it'll give you the same testbility.